### PR TITLE
Lasso delayed feedbacks algorithm

### PR DESCRIPTION
**Pull request by jdematias**
Lasso delayed feedbacks algorithm

### CODE REVIEW

**Comment by jdematias** [comment]
@VenturaDelMonte  
I know unit test is not completed but I would like that you give me your feedback about implementation

File: src/main/scala/eu/proteus/solma/lasso/LassoDFStreamTransformOperation.scala [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+class FitWindowFunction extends WindowFunction[LassoStreamEvent, OptionLabeledVector, Long, GlobalWindow] {
+
+  def apply(key: Long, window: GlobalWindow, input: Iterable[LassoStreamEvent],
+            out: Collector[OptionLabeledVector]): Unit =
+  {
+    val inputStreamEvents = input.filter(x => x.isLeft).map(x => x.left.get)
+    val labelStreamEvent = input.filter(x => x.isRight).map(x => x.right.get).head
+
+    val poses = inputStreamEvents.toList.map(x => x.pos._2)
+    var labels = List[(Double, Double)]()
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/lasso/LassoDFStreamTransformOperation.scala]: im not sure a List is the cheapest data structure to use here.

File: src/main/scala/eu/proteus/solma/lasso/LassoDFStreamTransformOperation.scala [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      def selectKey(event: LassoStreamEvent): Long = {
+        event match {
+          case Left(ev) => ev.pos._1
+          case Right(ev) => ev.label
+        }
+      }
+      val processedInput = rawInput.keyBy(x => selectKey(x)).window(GlobalWindows.create())
+        .trigger(new FitTrigger).apply(new FitWindowFunction())
+
+      val output = LassoParameterServer.transformLasso(None)(processedInput, workerParallelism = 3,
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/lasso/LassoDFStreamTransformOperation.scala]: why do you bound the parallelism to 3? Wouldnt it be better to extract the parallelism from the execution environment?

Comment (VenturaDelMonte) [reply src/main/scala/eu/proteus/solma/lasso/LassoDFStreamTransformOperation.scala]: Moreover, there are a lot of hard coded params. This is ok at this stage, imho. However, we might need to rethink this part later on (or generalize those settings now).

File: src/main/scala/eu/proteus/solma/lasso/algorithm/FlatnessMappingAlgorithm.scala [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ * limitations under the License.
+ */
+
+package eu.proteus.solma.lasso.algorithm
+
+/**
+  * TODO description
+  *
+  */
+class FlatnessMappingAlgorithm(input: List[Double],
```
Comment (VenturaDelMonte) [inline src/main/scala/eu/proteus/solma/lasso/algorithm/FlatnessMappingAlgorithm.scala]: I would suggest to check if the data structures used here are cheap enough w.r.t to the workload (e.g., index based lookups cost O(n) for a Scala List).

**Comment by jdematias** [comment]
@VenturaDelMonte I have taken your suggestions into account and I have changed PR, but It's not totally finished. I have problem simulating real streaming working. I have two streams, one with features measurements and other with flatness measurements which are joined in a unique stream. The problem is that flatness should be processed always after all features measurements have been processed (like work in the real factory). Now I have added a "bad trigger" (trigger throw calculation when 70 events (a hardcoded value) are received) only to test the algorithm, but obviously this can not be the final solution. Could you suggest me a better way to simulate it?

**Comment by VenturaDelMonte** [comment]
@jdematias I think the "custom trigger" you used to simulate is ok. An other solution would be to perform a join through a coMap/coFlatMap that would ingest both streams (say stream A containing "common" measurements and stream B containing flatness measurement). When items arrive to the operator  from stream A, they are immediately flushed downstream. Items arriving from stream B, they are cached. When measurements for a given coil are over, the cached flatness measurement for that coil is also sent downstream. We could track the ending of a coil through detecting a change of ids (we assume coils are produced in a FIFO fashion, right?). I believe this is a quick way to simulate our real scenario. What do you think about that?

**Comment by jdematias** [comment]
Thanks @VenturaDelMonte !! It seems a good idea. I will try to code it and add new changes to PR

**Comment by jdematias** [comment]
Hi @VenturaDelMonte
PR is OK for me. Please review it as soon as possible.
Thanks!!

**Comment by VenturaDelMonte** [comment]
Great job! Merging and closing.

**Pull request merged**